src/watchdog/src/watchdog.py

```diff
@@ -185,35 +190,37 @@ def __init__(self, name, gpu):
     def __repr__(self):
         return "%s: %s" % (self.name, self.gpu)
 
-def parse_pod_item(pod, pai_pod_gauge, pai_container_gauge, pods_info):
-    """ add metrics to pai_pod_gauge or pai_container_gauge if successfully paesed pod.
+def parse_pod_item(pod, pai_pod_gauge, pai_container_gauge, pai_job_pod_gauge, pods_info):
+    """ add metrics to pai_pod_gauge or pai_container_gauge if successfully parse pod.
     Because we are parsing json outputed by k8s, its format is subjected to change,
     we should test if field exists before accessing it to avoid KeyError """
 
     pod_name = pod["metadata"]["name"]
     namespace = walk_json_field_safe(pod, "metadata", "namespace") or "default"
     host_ip = walk_json_field_safe(pod, "status", "hostIP") or "unscheduled"
-
-    used_gpu = 0
+    status = pod["status"]
     containers = walk_json_field_safe(pod, "spec", "containers")
-    if containers is not None:
-        for container in containers:
-            req_gpu = int(walk_json_field_safe(container, "resources", "requests",
-                    "nvidia.com/gpu") or 0)
-            limit_gpu = int(walk_json_field_safe(container, "resources", "limits",
-                    "nvidia.com/gpu") or 0)
-            used_gpu += max(req_gpu, limit_gpu)
-    pods_info[host_ip].append(PodInfo(pod_name, used_gpu))
-
     labels = pod["metadata"].get("labels")
-    if labels is None or "app" not in labels:
-        logger.info("unknown pod %s", pod["metadata"]["name"])
+
+    service_name = labels['app'] if labels is not None and 'app' in labels else None
+    job_name = labels['jobName'] if labels is not None and 'jobName' in labels else None
+    if service_name is None and job_name is None:
+        logger.info("unknown pod %s", pod_name)
         return None
 
-    service_name = labels["app"] # get pai service name from label
+    generate_pods_info(pod_name, containers, host_ip, pods_info)
+    generate_pod_metrics(pai_pod_gauge, pai_job_pod_gauge,
+                         service_name, job_name, pod_name, host_ip, status, namespace)
 
-    status = pod["status"]
+    # generate pai_containers
+    if service_name is not None and status.get("containerStatuses") is not None:
+        container_statuses = status["containerStatuses"]
+        generate_container_metrics(
+            pai_container_gauge, service_name, pod_name, container_statuses, namespace, host_ip)
 
+
+def generate_pod_metrics(pai_pod_gauge, pai_job_pod_gauge, service_name, job_name, pod_name,
+                         host_ip, status, namespace):
```

Comment:
Maybe add `assert service_name is not None and job_name is not None` to make sure caller passed expected parameter

Reply:
Here we parse all pods running in k8s. Some pods such as nvidia-device-plugin will not contain app label and job label.

---
In reply to: [337403770](https://github.com/microsoft/pai/pull/3761#discussion_r337403770) [](ancestors = 337403770)

Reply:
No, you returned in caller in this case. Sorry, should be `assert service_name is not None or job_name is not None`

Reply:
This function only generate metrics to job and service, the caller must know this, assert statement is to let caller know it, so I suggest adding this so in future, code reader or writer can know this by only looking assert